core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -322,31 +365,25 @@ protected void validateDataFilesExist(TableMetadata base, Long startingSnapshotI
 
       if (matchingOperations.contains(currentSnapshot.operation())) {
         newSnapshots.add(currentSnapshotId);
-        for (ManifestFile manifest : currentSnapshot.dataManifests()) {
-          if (manifest.snapshotId() == (long) currentSnapshotId) {
-            manifests.add(manifest);
+        if (content == ManifestContent.DATA) {
+          for (ManifestFile manifest : currentSnapshot.dataManifests()) {
+            if (manifest.snapshotId() == (long) currentSnapshotId) {
```

Comment:
Is it possible that the `manifest.snapshotId()` returns a `null` value  ? If so, then we'd better to change this line to `Objects.equals(manifest.snapshotId(), currentSnapshotId)` ?

Reply:
No, there should never be a null snapshot ID returned from `ManifestFile`. `null` was allowed because manifest lists were added later, but manifest metadata is read any time a table is written so that the manifest data is reliable.